proto/substrait/extended_expression.proto

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: Apache-2.0
+syntax = "proto3";
+
+package substrait;
+
+import "substrait/algebra.proto";
+import "substrait/extensions/extensions.proto";
+import "substrait/plan.proto";
+import "substrait/type.proto";
+
+option csharp_namespace = "Substrait.Protobuf";
+option go_package = "github.com/substrait-io/substrait-go/proto";
+option java_multiple_files = true;
+option java_package = "io.substrait.proto";
+
+message ExpressionReference {
+  oneof expr_type {
+    Expression expression = 1;
+    AggregateFunction measure = 2;
+  }
+  // Field names in depth-first order
+  repeated string output_names = 3;
+}
+
+// Describe a set of operations to complete.
+// For compactness sake, identifiers are normalized at the plan level.
+message ExtendedExpression {
+  // Substrait version of the expression. Optional up to 0.17.0, required for later
+  // versions.
+  Version version = 7;
+
+  // a list of yaml specifications this expression may depend on
+  repeated substrait.extensions.SimpleExtensionURI extension_uris = 1;
+
+  // a list of extensions this expression may depend on
+  repeated substrait.extensions.SimpleExtensionDeclaration extensions = 2;
+
+  // one or more expression trees with same order in plan rel
+  repeated ExpressionReference referred_expr = 3;
```

Comment:
I still find it confusing that we have a repeated list of expressions here.  I think it would be more intuitive to have:

 * One top level expression
 * A repeated list of sub-expressions

We could do something similar to what we do with plans:

```
// An expression with output field names.
//
// This is for use at the root of an `Expression` tree.
message ExpressionRoot {
  // The root expression
  Expression expression = 1;
  // Field names in depth-first order
  repeated string names = 2;
}

message StandaloneExpression {
  oneof rel_type {
    // A sub-expression (used for references)
    Expression expr = 1;
    // The root of an expression tree
    ExpressionRoot root = 2;
  }
}

message ExtendedExpression {
  ...
  repeated StandaloneExpression exprs;
  ...
}
```

However, I think I would prefer something more straightforward:

```
message ExtendedExpression {
...
// Top level expression
Expression expr = 3;
repeated string output_field_names = 4;
// Common sub-expressions (referenced by `expr`)
repeated Expression sub_expressions = 5;
...
}
```

Although, if we end up adopting something like https://github.com/substrait-io/substrait/issues/415 then we could use repeated sub-expressions everywhere and we wouldn't need any of this.  It could just be:

```
message ExtendedExpression {
...
// Top level expression (may contain let expressions)
Expression expr = 3;
repeated string output_field_names = 4;
...
}
```

Reply:
Thanks for your examples. I think the greatest benefit is that a `let` expression is explicitly highlighted so that consumers doesn't need expr analysis again for an optimization like "CSE". This is a good idea but it should be contained in message `Expression` inside? If so, `ExtendedExpression` can keep as current.

Reply:
If we had `let` expression then we do not need the top level expression to be repeated.

However, I agree that the current proposal will still work if we introduce `let` expressions.

Reply:
`let` expression will make CSE easier but it also need producer does this expression level optimization and stores that info before translation to substrait. We are happy to do refine on this `ExtendedExpression` message when this idea finishes adoption.